contracts/AssetPool.sol

```diff
@@ -80,17 +124,112 @@ contract AssetPool is Ownable {
             "Underwriter token amount must be greater than 0"
         );
 
+        pendingWithdrawal[msg.sender] = covAmount.add(
+            pendingWithdrawal[msg.sender]
+        );
+        /* solhint-disable not-rely-on-time */
+        withdrawalInitiatedTimestamp[msg.sender] = block.timestamp;
+
+        emit WithdrawalInitiated(
+            msg.sender,
+            pendingWithdrawal[msg.sender],
+            block.timestamp
+        );
+
+        underwriterToken.safeTransferFrom(msg.sender, address(this), covAmount);
+        /* solhint-enable not-rely-on-time */
+    }
+
+    /// @notice Completes the previously initiated withdrawal for the
+    ///         underwriter. Anyone can complete the withdrawal for the
+    ///         underwriter who previously initiated it.
+    ///         Depending on how long it took to complete the withdrawal since
+    ///         the time it has been initiated, part of the collateral and
+    ///         rewards can be seized by the pool.
+    ///         After the withdrawal delay and below the graceful withdrawal
+    ///         timeout, no tokens are seized by the pool. After the graceful
+    ///         withdrawal timeout, tokens are slowly getting seized by the pool
+    ///         over time. After the hard withdrawal timeout, 99% of tokens is
+    ///         seized by the pool and 1% of tokens is sent to the notifier who
+    ///         completed the withdrawal on behalf of the underwriter.
+    function completeWithdrawal(address underwriter) external {
+        /* solhint-disable not-rely-on-time */
+        uint256 initiatedAt = withdrawalInitiatedTimestamp[underwriter];
+        require(initiatedAt > 0, "No withdrawal initiated for the underwriter");
+
+        uint256 withdrawalDelayEndTimestamp = initiatedAt.add(withdrawalDelay);
+        require(
+            withdrawalDelayEndTimestamp < block.timestamp,
+            "Withdrawal delay has not elapsed"
+        );
+
+        // slither-disable-next-line reentrancy-events
         rewardsPool.withdraw();
 
+        uint256 covAmount = pendingWithdrawal[underwriter];
         uint256 covSupply = underwriterToken.totalSupply();
         uint256 collateralBalance = collateralToken.balanceOf(address(this));
 
         uint256 amountToWithdraw =
             covAmount.mul(collateralBalance).div(covSupply);
 
-        underwriterToken.safeTransferFrom(msg.sender, address(this), covAmount);
+        //
+        //      withdrawal              graceful withdrawal
+        //        delay                 timeout
+        //  /--------------\ /----------|
+        // x----------------x-----------x-------------------------------x------>
+        // ^                 \-----------------------------------------|
```

Comment:
it looks like the hard withdrawal timeout starts counting when withdrawal delay ends, but in the comment for the hard withdrawal it says that the time starts ticking at the moment of initiation. Can we maybe start the line from `initiatedAt` moment for the `hard withdrawal timeout`?

Reply:
It starts counting after the withdrawal delay has passed, that comment is wrong, I will fix it.